test/hotspot/jtreg/compiler/intrinsics/string/TestCountPositives.java

```diff
@@ -21,88 +21,121 @@
-        int len, off;
-        int ng;
+        for (int off = 0; off < 16; off++) { // starting offset of array segment
+            // Test all array segment sizes 1-63
+            for (int len = 1; len < 64; len++) {
+                test_countPositives(off, len, 0, 0);
+                test_countPositives(off, len, 1, 0);
+                test_countPositives(off, len, RANDOM.nextInt(30) + 2, 0);
+            }
+            // Test a random selection of sizes between 64 and 4099, inclusive
+            for (int i = 0; i < 20; i++) {
+                int len = 64 + RANDOM.nextInt(4100 - 64);
+                test_countPositives(off, len, 0, 0);
+                test_countPositives(off, len, 1, 0);
+                test_countPositives(off, len, RANDOM.nextInt(len) + 2, 0);
+            }
+            for (int len : new int[] { 128, 2048 }) {
+                // test with negatives only in a 1-63 byte tail
+                int tail = RANDOM.nextInt(63) + 1;
+                int ng = RANDOM.nextInt(tail) + 1;
+                test_countPositives(off, len + tail, ng, len);
+            }
+        }
+    }
 
-        for (ng = 0; ng < 57; ++ng) { // number of negatives in array segment
-            for (off = 0; off < 8; ++off) { // starting offset of array segment
-                for (int i = 0; i < sizes.length; ++i) { // array segment size
-                                                         // choice
-                    len = sizes[i];
-                    if (len + off > tBa.length)
-                        continue;
-                    initialize(off, len, ng);
-                    int calculated = Helper.StringCodingCountPositives(tBa, off, len);
-                    int expected = countPositives(tBa, off, len);
-                    if (calculated != expected) {
-                        if (expected != len && calculated >= 0 && calculated < expected) {
-                            // allow intrinsics to return early with a lower value,
-                            // but only if we're not expecting the full length (no
-                            // negative bytes)
-                            continue;
-                        }
-                        throw new Exception("Failed test countPositives " + "offset: " + off + " "
-                                + "length: " + len + " " + "return: " + calculated + " expected: " + expected + " negatives: "
-                                + ng);
-                    }
-                }
+    private static void test_countPositives(int off, int len, int ng, int ngOffset) throws Exception {
+        assert (len + off < bytes.length);
+        initialize(off, len, ng, ngOffset);
+        int calculated = Helper.StringCodingCountPositives(bytes, off, len);
+        int expected = countPositives(bytes, off, len);
+        if (calculated != expected) {
+            if (expected != len && ng >= 0 && calculated >= 0 && calculated < expected) {
```

Comment:
Why does this happen? Why does `countPositives` not get the same result here?

Reply:
The `countPositives` intrinsics were adapted from the pre-existing `hasNegatives` intrinsic, and to ease porting over all platforms - including those we don't maintain - we specified `countPositives` so that if there's a negative value at index `idx` then it's acceptable for the intrinsics to return any value from `0` to `idx - 1`. This way intrinsics on all platforms could quickly be adapted over (return either `0` for `hasNegatives == true` or `length` for `hasNegatives == false`) while respective maintainers worked on porting over properly. 

This also allows the intrinsic to do a quick approximation: E.g. if a vectorized loop has checked 128 bytes but then detects a negative in the next block then the intrinsic can opt to return 128 rather than whatever the exact count. The surrounding java code will then do a scalar loop that is roughly as fast as an intrinsic tail routine would be, and with less complexity. I believe the `aarch64` intrinsic is implemented like this.

Reply:
Ok, then you could add such a comment to the code. Optional.

Reply:
I think it's roughly covered by the `@implSpec`:
```
    /**
     * Count the number of leading positive bytes in the range.
     *
     * @implSpec the implementation must return len if there are no negative
     *   bytes in the range. If there are negative bytes, the implementation must return
     *   a value that is less than or equal to the index of the first negative byte
     *   in the range.
     */
    @IntrinsicCandidate
    public static int countPositives(byte[] ba, int off, int len) {
```
.. but of course the historical and current rationale behind that specification might not be immediately clear.